packages/sdk/src/protocols/filter.ts

```diff
@@ -57,28 +62,37 @@ export class SubscriptionManager {
   async subscribe<T extends IDecodedMessage>(
     decoders: IDecoder<T> | IDecoder<T>[],
     callback: Callback<T>
-  ): Promise<void> {
+  ): Promise<SDKProtocolResult> {
     const decodersArray = Array.isArray(decoders) ? decoders : [decoders];
 
     // check that all decoders are configured for the same pubsub topic as this subscription
-    decodersArray.forEach((decoder) => {
+    for (const decoder of decodersArray) {
       if (decoder.pubsubTopic !== this.pubsubTopic) {
-        throw new Error(
-          `Pubsub topic not configured: decoder is configured for pubsub topic ${decoder.pubsubTopic} but this subscription is for pubsub topic ${this.pubsubTopic}. Please create a new Subscription for the different pubsub topic.`
-        );
+        return {
+          failures: [
+            {
+              error: ProtocolError.TOPIC_DECODER_MISMATCH
+            }
+          ],
+          successes: []
```

Comment:
I'm concerned with how much implicit state to handle is being passed on to the library consumer. Right now the `SDKProtocolResult` can mean several different things:
1. the state here, when there is a single error in `failures` and no item in `successes`, which is like a "total failure"
2. some items in `successes` and some items in `failures`, which means some subscriptions were successful
3. some items in `successes` and no items in `failures`, which means all subscriptions were successful
4. no items in `successes` and no items in `failures`, which _should_ not happen, but technically is possible according to the type

It would be better if this function could only ever return either a failure or a success. But that might mean limiting it to one decoder at a time.

Reply:
> It would be better if this function could only ever return either a failure or a success. But that might mean limiting it to one decoder at a time.

Correct. `nwaku`/protobuf currently allows us to pass multiple content topics to the **same** request -- limiting to one decoder (->one content topic) on js-waku API side will be limiting of that.
This seems to be an API change that we will introduce.

I agree with your overall comment of the potential mindload on library consumer, as it could be confusing to have both `failure` and `success` haha.

Not sure what the right step is. 
We can track this in an issue and discuss them in the next PM call @adklempner

Reply:
I created an issue for it, let me know if I captured it correctly: https://github.com/waku-org/js-waku/issues/1981